base/lttagging.dtx

```diff
@@ -0,0 +1,789 @@
+% in this file.
+%
+% This file is part of the LaTeX base system.
+% -------------------------------------------
+%
+% It may be distributed and/or modified under the
+% conditions of the LaTeX Project Public License, either version 1.3c
+% of this license or (at your option) any later version.
+% The latest version of this license is in
+%    https://www.latex-project.org/lppl.txt
+% and version 1.3c or later is part of all distributions of LaTeX
+% version 2008 or later.
+%
+% This file has the LPPL maintenance status "maintained".
+%
+% The list of all files belonging to the LaTeX base distribution is
+% given in the file `manifest.txt'. See also `legal.txt' for additional
+% information.
+%
+% The list of derived (unpacked) files belonging to the distribution
+% and covered by LPPL is defined by the unpacking scripts (with
+% extension .ins) which are part of the distribution.
+%
+% \fi
+%
+% \iffalse
+%%% From File: lttagging.dtx
+%
+%<*driver>
+% \fi
+\ProvidesFile{lttagging.dtx}
+             [2023/12/01 v1.0a LaTeX Kernel (tagging support)]
+% \iffalse
+\documentclass{l3doc}
+\GetFileInfo{lttagging.dtx}
+\title{\filename}
+\date{\filedate}
+\author{\LaTeX{} project}
+
+\begin{document}
+% \MaintainedByLaTeXTeam{latex}     % should be added again the moment
+                                    % it is supported by l3doc
+ \maketitle
+ \DocInput{\filename}
+\end{document}
+%</driver>
+% \fi
+%
+% \providecommand\env[1]{\texttt{#1}}
+%
+% \providecommand\hook[1]{\texttt{#1\DescribeHook[noprint]{#1}}}
+% \providecommand\socket[1]{\texttt{#1\DescribeSocket[noprint]{#1}}}
+% \providecommand\plug[1]{\texttt{#1\DescribePlug[noprint]{#1}}}
+%
+% \let\ProvideDocElement\NewDocElement
+%
+% \ProvideDocElement[printtype=\textit{socket},idxtype=socket,idxgroup=Sockets]{Socket}{socketdecl}
+% \ProvideDocElement[printtype=\textit{hook},idxtype=hook,idxgroup=Hooks]{Hook}{hookdecl}
+% \ProvideDocElement[printtype=\textit{plug},idxtype=plug,idxgroup=Plugs]{Plug}{plugdecl}
+%
+%
+% \section{}
+%
+%
+%
+% \MaybeStop{}
+%
+%    \begin{macrocode}
+%<*2ekernel|latexrelease>
+%    \end{macrocode}
+%
+%    \begin{macrocode}
+\ExplSyntaxOn
+%    \end{macrocode}
+%
+%
+% \section{General support for tagged output}
+%
+%  \DescribeMacro\SuspendTagging
+%  \DescribeMacro\ResumeTagging
+%  The are places in code where it is import top stop any tagging
+%  activities, e.g., when we are doing trial typesetting that it is done
+%  several times. In such a case one must tag only the final version
+%  that is actually used, otherwise tagging structures are allowed
+%  which then do not end up in the PDF and confuse the mechanism. For
+%  this we have two commands that can be used in packages:
+%  \cs{SuspendTagging} and \cs{ResumeTagging}. They are available as
+%  part of the \LaTeX{} kernel, so that they can be safely used in
+%  packages whether or not tagging is requested. By default they do
+%  nothing.
+%
```

Comment:
Currently tagging should not be stopped while the output routine is called (as header/footer/footnoterule etc should always be tagged as artifact). We probably will have to guard the output routine, but for now this restriction should at least be documented.

Reply:
From a conceptional point of view tagging inside the OR is independent of tagging in the document, e.g., the OR has to establish its on tagging --- the interface  needs to be decided but it has to be available. The restriction as you call it isn't really a restriction that the user can control unless everything is inside a box, so rather than to document it we have to resolve it soon

Reply:
yes I agree that the OR has to guard itself. On the other side, suspending  (completly stopping) tagging should imho be done only in small, controlled places, mostly if boxes are involved e.g. for measuring.

Reply:
agreed, but this is why in a new context (OR and possibly also floats) tagging should be re-established if it was active before  \SuspendTagging was executed. So it needs 2 control variables, one for the overall status (tagging=yes/no) and one for temporary suspension.

Reply:
as I used now an int to keep track of start/stop one can use that to check for temporary suspension. Which means I should release the next tagpdf version ;-)

Reply:
yes but only if it includes proper definitions for `\SuspendTagging` and friends so we should be sure first that we want those with these names etc. Then we can drop the redefinitions in lttagging and only provide the dummies.